pkg/relayer/session/session_test.go

```diff
@@ -1,3 +1,108 @@
 package session_test
 
-// TODO: Add tests to the relayerSessionsManager logic
+import (
+	"context"
+	"testing"
+	"time"
+
+	"cosmossdk.io/depinject"
+	"github.com/stretchr/testify/require"
+
+	"github.com/pokt-network/poktroll/pkg/client"
+	"github.com/pokt-network/poktroll/pkg/observable/channel"
+	"github.com/pokt-network/poktroll/pkg/relayer"
+	"github.com/pokt-network/poktroll/pkg/relayer/miner"
+	"github.com/pokt-network/poktroll/pkg/relayer/session"
+	"github.com/pokt-network/poktroll/testutil/testclient/testblock"
+	"github.com/pokt-network/poktroll/testutil/testclient/testsupplier"
+	"github.com/pokt-network/poktroll/testutil/testrelayer"
+	servicetypes "github.com/pokt-network/poktroll/x/service/types"
+	sessiontypes "github.com/pokt-network/poktroll/x/session/types"
+)
+
+func TestRelayerSessionsManager_Start(t *testing.T) {
+	const (
+		sessionStartHeight = 1
+		sessionEndHeight   = 2
+	)
+	var (
+		zeroByteSlice = []byte{0}
+		ctx           = context.Background()
+	)
+
+	// Set up dependencies.
+	blocksObs, blockPublishCh := channel.NewReplayObservable[client.Block](ctx, 1)
+	blockClient := testblock.NewAnyTimesCommittedBlocksSequenceBlockClient(t, blocksObs)
+	supplierClient := testsupplier.NewOneTimeClaimProofSupplierClient(ctx, t)
+
+	deps := depinject.Supply(blockClient, supplierClient)
+	storesDirectoryOpt := testrelayer.WithTempStoresDirectory(t)
+
+	// Create a new relayer sessions manager.
+	relayerSessionsManager, err := session.NewRelayerSessions(ctx, deps, storesDirectoryOpt)
+	require.NoError(t, err)
+	require.NotNil(t, relayerSessionsManager)
+
+	// Pass a mined relays observable via InsertRelays.
+	mrObs, minedRelaysPublishCh := channel.NewObservable[*relayer.MinedRelay]()
+	minedRelaysObs := relayer.MinedRelaysObservable(mrObs)
+	relayerSessionsManager.InsertRelays(minedRelaysObs)
+
+	// Start the relayer sessions manager.
+	relayerSessionsManager.Start(ctx)
+
+	// Publish a mined relay to the minedRelaysPublishCh to insert into the session tree.
+	minedRelay := newMinedRelay(t, sessionStartHeight, sessionEndHeight)
+	minedRelaysPublishCh <- minedRelay
+
+	// Wait a tick to allow the relayer sessions manager to process asynchronously.
+	// It should have created a session tree for the relay.
```

Comment:
1. We could already make assertions here by checking the filesystem. I was planning on doing that in a separate test which focuses on correctness of persisted session trees.
2. The shortest path I see to a refactor which allows for unit-level isolation of these parts would be to refactor `#ensureSessionTree()`  such that the call to `NewSessionTree()` is instead referencing a function assigned to a field on `relayerSessionsManager`. This function could then be overridden via an option that's passed to `NewRelayerSessions()`. Since `SessionTree` is an interface (:mechanical_arm:), in tests, this function can return a mock of the `SessionTree`.

Reply:
> We could already make assertions here by checking the filesystem.

Given that we have the optionality of updating the interface, I think that's cleaner.

> The shortest path I see to a refactor which allows  ...

I'd want to see what that looks like (to make sure it doesn't overcomplicate things), but makes sense overall.

I was going to just `RelaySessionManager` getting a `WithMocks` config option, but then we're introducing test logic into the core. Might be simpler code, but I wouldn't be able to argue against the solution you proposed above.